In [ ]:
# pip install python-dotenv pandas os openpyxl

## Overview of Data

Show the distribution of all original files

In [10]:
import os
from dotenv import load_dotenv
import pandas as pd

# load_dotenv(dotenv_path=".env")

folder_path = "./data"

if not folder_path:
    raise ValueError("path not found")

files = [f for f in os.listdir(folder_path) if f.endswith('.xlsx')]

for file in sorted(files):
    path = os.path.join(folder_path, file)
    df = pd.read_excel(path)
    print(f"{file}: {df.shape[0]} rows × {df.shape[1]} columns")

2025-01-01.xlsx: 4296 rows × 63 columns
2025-01-08.xlsx: 4296 rows × 63 columns
2025-01-15.xlsx: 4296 rows × 63 columns
2025-01-22.xlsx: 4296 rows × 63 columns
2025-01-29.xlsx: 4296 rows × 63 columns
2025-02-05.xlsx: 4296 rows × 63 columns
2025-02-12.xlsx: 4296 rows × 63 columns
2025-02-19.xlsx: 4301 rows × 63 columns
2025-02-26.xlsx: 4301 rows × 63 columns
2025-03-05.xlsx: 4301 rows × 63 columns
2025-03-12.xlsx: 4301 rows × 63 columns
2025-03-19.xlsx: 4301 rows × 63 columns
2025-03-26.xlsx: 4301 rows × 74 columns
2025-04-02.xlsx: 4301 rows × 74 columns
2025-04-09.xlsx: 4301 rows × 74 columns
2025-04-16.xlsx: 4301 rows × 74 columns
2025-04-23.xlsx: 4301 rows × 74 columns
2025-04-30.xlsx: 4301 rows × 74 columns
2025-05-07.xlsx: 4301 rows × 74 columns
2025-05-14.xlsx: 4301 rows × 74 columns
2025-05-21.xlsx: 4301 rows × 74 columns
2025-05-28.xlsx: 4301 rows × 74 columns
2025-06-04.xlsx: 4301 rows × 74 columns
2025-06-11.xlsx: 4301 rows × 74 columns
2025-06-18.xlsx: 4301 rows × 74 columns


There is a variation in number of columns (63 -74) and number of rows (4296 - 4301 -4300)

In [18]:
file_old = './data/2025-03-19.xlsx'
file_new = './data/2025-03-26.xlsx'

cols_old = pd.read_excel(file_old, header=4, nrows=0).columns
cols_new = pd.read_excel(file_new, header=4, nrows=0).columns

added = [col for col in cols_new if col not in cols_old]
removed = [col for col in cols_old if col not in cols_new]
common = [col for col in cols_new if col in cols_old]

print(f"In common {len(common)}")
print(f"New cols in {file_new}:")
for col in added:
    print(f"  + {col}")

print(f"\nDeleted cols in {file_old}:")
for col in removed:
    print(f"  - {col}")


In common 50
New cols in ./data/2025-03-26.xlsx:
  + USMCA Eligible Flag
  + Open Orders GDRC-D
  + Open WO GDRC-D
  + Open WO GDRC-H
  + For Sterilization GDRC-D
  + For Sterilization GDRC-H
  + Intransit to GDRC -H
  + GDRC-H FG 7-9 MO
  + ADD GDRC-D Backorder
  + Overlay GDRC-D(For Information Only)
  + Overlay GDRC-H(For Information Only))
  + Overlay ADD GDRC-D
  + Overlay ADD GDRC-H
  + Net Available Inventory GDRC-D
  + Net Available Inventory GDRC-H
  + Net Available Inventory Total
  + Target Inventory GDRC-D
  + Target Inventory GDRC-H
  + Target Inventory Total
  + Days on Hand GDRC-D
  + Days on Hand GDRC-H
  + Days On Hand Total
  + RR Suggested Order GDRC-D(For Information Only)
  + RR Suggested Order LSP (For Information Only)

Deleted cols in ./data/2025-03-19.xlsx:
  - USMCA_Eligible_Flag
  - Open Orders GDRC D
  - Open WO
  - For Sterilization
  - ADD GDRC D Backorder
  - Overlay (For Information Only)
  - Overlay ADD-H
  - Overlay ADD-D
  - Overlay ADD
  - Net Availa

In [ ]:
file_old = './data/2025-03-19.xlsx'
file_new = './data/2025-03-26.xlsx'

cols_old = pd.read_excel(file_old, header=4, nrows=0).columns
cols_new = pd.read_excel(file_new, header=4, nrows=0).columns

added = [col for col in cols_new if col not in cols_old]
removed = [col for col in cols_old if col not in cols_new]
common = [col for col in cols_new if col in cols_old]

print(f"In common {len(common)}")
print(f"New cols in {file_new}:")
for col in added:
    print(f"  + {col}")

print(f"\nDeleted cols in {file_old}:")
for col in removed:
    print(f"  - {col}")


In common 50
New cols in ./data/2025-03-26.xlsx:
  + USMCA Eligible Flag
  + Open Orders GDRC-D
  + Open WO GDRC-D
  + Open WO GDRC-H
  + For Sterilization GDRC-D
  + For Sterilization GDRC-H
  + Intransit to GDRC -H
  + GDRC-H FG 7-9 MO
  + ADD GDRC-D Backorder
  + Overlay GDRC-D(For Information Only)
  + Overlay GDRC-H(For Information Only))
  + Overlay ADD GDRC-D
  + Overlay ADD GDRC-H
  + Net Available Inventory GDRC-D
  + Net Available Inventory GDRC-H
  + Net Available Inventory Total
  + Target Inventory GDRC-D
  + Target Inventory GDRC-H
  + Target Inventory Total
  + Days on Hand GDRC-D
  + Days on Hand GDRC-H
  + Days On Hand Total
  + RR Suggested Order GDRC-D(For Information Only)
  + RR Suggested Order LSP (For Information Only)

Deleted cols in ./data/2025-03-19.xlsx:
  - USMCA_Eligible_Flag
  - Open Orders GDRC D
  - Open WO
  - For Sterilization
  - ADD GDRC D Backorder
  - Overlay (For Information Only)
  - Overlay ADD-H
  - Overlay ADD-D
  - Overlay ADD
  - Net Availa

New REAL columns:

USMCA_Eligible_Flag --> USMCA Eligible Flag

Open Orders GDRC D --> Open Orders GDRC-D

Open WO --> Open WO GDRC-D //
            Open WO GDRC-H

For Sterilization --> For Sterilization GDRC-D //
                      For Sterilization GDRC-H

D + --> Intransit to GDRC -H

D + --> GDRC-H FG 7-9 MO

ADD GDRC D Backorder --> ADD GDRC-D Backorder
  - Overlay (For Information Only)
  - Overlay ADD-H
  - Overlay ADD-D
  - Overlay ADD

    -->
  + Overlay GDRC-D(For Information Only)
  + Overlay GDRC-H(For Information Only))
  + Overlay ADD GDRC-D
  + Overlay ADD GDRC-H

Net Available Inventory --> Net Available Inventory GDRC-D //
                            Net Available Inventory GDRC-H //
                            Net Available Inventory Total

Target Inventory --> Target Inventory GDRC-D //
                     Target Inventory GDRC-H //
                     Target Inventory Total

Days on Hand --> Days on Hand GDRC-D //
                 Days on Hand GDRC-H //
                 Days On Hand Total

RR Suggested Order(For Information Only) --> RR Suggested Order GDRC-D(For Information Only) //
                                             RR Suggested Order LSP (For Information Only)

## Mapping Step + Concatenate Files:
- Define custom rules based on how we want to manage the updates - column mapping
- Define selected columns
- Create a function to apply these rules to the selected colums
- Concatenate prepared files

We have created a "test" folder for using some files for testing

In [ ]:
# 1. Define column mapping
# -----------------------------
COLUMN_MAP = {
    'USMCA_Eligible_Flag': 'USMCA Eligible Flag',
    'Net Available Inventory': 'Net Available Inventory Total',
    'Open Orders GDRC D': 'Open Orders GDRC-D',
    'ADD GDRC D Backorder': 'ADD GDRC-D Backorder',
    

}

# -----------------------------
# 2. Column selection for modeling
# -----------------------------
SELECTED_COLUMNS = [
    'Product ID', 'Brand', 'MFG Class', 'Resource ID', 'Franchise View', 'USMCA Eligible Flag', 'FDA FTZ Flag',
    'ABC Classification', 'Backorder/Backlog Total', 'Open Orders Total',
    'Net Available Inventory Total', 'LSP DOH Target', 'US DOH Target', 'Std Cost', 'ADD Total Incl. backorder'
]

# -----------------------------
# 3. Map function to unify column names
# -----------------------------
def unify_column_name(col):
    return COLUMN_MAP.get(col.strip(), col.strip())

# -----------------------------
# 4. Function to load and unify one file
# -----------------------------
def load_and_unify_excel(path):
    df = pd.read_excel(path, header=4)  # Asumes encabezado en fila 5
    df.rename(columns=lambda col: unify_column_name(col), inplace=True)
    df = df.dropna(how='all')  # Eliminar filas completamente vacías

    # Eliminar filas donde no hay Product ID y todas las columnas seleccionadas están vacías o nulas
    condition = df[SELECTED_COLUMNS].isna().all(axis=1)
    if 'Product ID' in df.columns:
        condition &= df['Product ID'].isna()
    df = df[~condition]

    for col in SELECTED_COLUMNS:
        if col not in df.columns:
            df[col] = pd.NA  # Rellenar con NA si falta
    return df[SELECTED_COLUMNS]

# -----------------------------
# 5. Main script to concatenate
# -----------------------------
files_path = "./test"

def concatenate_psr_files(files_path):
    dfs = []
    for file in sorted(os.listdir(files_path)):
        if file.endswith('.xlsx'):
            full_path = os.path.join(files_path, file)
            df = load_and_unify_excel(full_path)
            df['week'] = os.path.splitext(file)[0]
            dfs.append(df)
    final_df = pd.concat(dfs, ignore_index=True)
    final_df['week'] = pd.to_datetime(final_df['week'], errors='coerce')
    return final_df


In [39]:
concatenate_psr_files(files_path)

Product ID Franchise View          Brand USMCA Eligible Flag  \
0      12000210         Cardio      ARC ROYAL                   N   
1      12000235         Cardio      ARC ROYAL                   N   
2      12000236         Cardio      ARC ROYAL                   N   
3      12000237         Cardio      ARC ROYAL                   N   
4      12000249         Cardio      ARC ROYAL                   N   
...         ...            ...            ...                 ...   
8585   ZE-6-120         Cardio  IKAZUCHI ZERO                   N   
8586   ZE-6-200         Cardio  IKAZUCHI ZERO                   N   
8587   ZE-6-250         Cardio  IKAZUCHI ZERO                   N   
8588  ZE-8-100P         Cardio  IKAZUCHI ZERO                   N   
8589   ZE-8-120         Cardio  IKAZUCHI ZERO                   N   

     FDA FTZ Flag ABC Classification  Backorder/Backlog Total  \
0             FTZ                 C                       0.0   
1             FTZ                 C                       0.0   
2             FTZ                 C                       0.0   
3             FTZ                 C                       0.0   
4             FTZ                 C                       0.0   
...           ...                ...                      ...   
8585          FTZ                 C                       0.0   
8586          FTZ                 C                       0.0   
8587          FTZ                 C                       0.0   
8588          FTZ                 C                       0.0   
8589          FTZ                 C                       0.0   

      Open Orders Total  Net Available Inventory Total  LSP DOH Target  \
0                   0.0                            0.0             0.0   
1                   0.0                            0.0             0.0   
2                   0.0                            0.0             0.0   
3                   0.0                            0.0             0.0   
4                   0.0                            0.0             0.0   
...                 ...                            ...             ...   
8585                0.0                           74.0            23.0   
8586                0.0                           50.0            23.0   
8587                0.0                           36.0            23.0   
8588                0.0                           73.0            23.0   
8589                3.0                           41.0            23.0   

      US DOH Target  Std Cost       week  
0               0.0    1.8983 2025-03-19  
1               0.0    1.4544 2025-03-19  
2               0.0    1.7255 2025-03-19  
3               0.0    3.2436 2025-03-19  
4               0.0    6.4364 2025-03-19  
...             ...       ...        ...  
8585           32.0   52.1382 2025-03-26  
8586           32.0   55.8372 2025-03-26  
8587           32.0   55.8372 2025-03-26  
8588           32.0   52.8380 2025-03-26  
8589           32.0   51.8400 2025-03-26  

[8590 rows x 13 columns]

Files are saved on "results" folder

In [41]:
output_path = os.path.join("./results", 'psr_full_dataset.xlsx')
concatenate_psr_files(files_path).to_excel(output_path, index=False)